### ISSUE TITLE
LibEvHandler broken due to b25c3f3

### ISSUE THREAD

**Issue by audaki** [CopernicaMarketingSoftware/AMQP-CPP#428]
**Describe the bug**
LibEvHandler doesn't work anymore (simple code from README.md without timer)

**Expected behavior and actual behavior**
The easy examples in the tutorial with LibEvHandler break due to this change

**Comment by rwols** [comment]
Can you point me to the example in the readme that fails? And in what way?

**Comment by audaki** [comment]
Sure, I'm gonna send a complete compilable example that showcases the problem.

```
#include <amqpcpp.h>
#include <amqpcpp/libev.h>
#include <amqpcpp/linux_tcp.h>
#include <ev.h>

class MyHandler : public AMQP::LibEvHandler
{
public:
  MyHandler(struct ev_loop* loop) : AMQP::LibEvHandler(loop) {}

  virtual ~MyHandler() = default;
  void onAttached(AMQP::TcpConnection* connection) override
  {
    std::cout << "onAttached" << std::endl;
    TcpHandler::onAttached(connection);
  }
  void onProperties(AMQP::TcpConnection* connection, const AMQP::Table& server, AMQP::Table& client) override
  {
    std::cout << "onProperties" << std::endl;
    TcpHandler::onProperties(connection, server, client);
  }
  void onHeartbeat(AMQP::TcpConnection* connection) override
  {
    std::cout << "onHeartbeat" << std::endl;
    TcpHandler::onHeartbeat(connection);
  }
  bool onSecuring(AMQP::TcpConnection* connection, SSL* ssl) override
  {
    std::cout << "onSecuring" << std::endl;
    return TcpHandler::onSecuring(connection, ssl);
  }
  bool onSecured(AMQP::TcpConnection* connection, const SSL* ssl) override
  {
    std::cout << "onSecured" << std::endl;
    return TcpHandler::onSecured(connection, ssl);

    return true;
  }
  void onLost(AMQP::TcpConnection* connection) override
  {
    std::cout << "onLost" << std::endl;
    TcpHandler::onLost(connection);
  }
  virtual void onReady(AMQP::TcpConnection* connection) override
  {
    std::cout << "ready" << std::endl;
  }
  virtual void onClosed(AMQP::TcpConnection* connection) override
  {
    std::cout << "closed" << std::endl;
  }
  virtual void onError(AMQP::TcpConnection* connection, const char* message) override
  {
    std::cout << "error: " << message << std::endl;
  }
  virtual void onConnected(AMQP::TcpConnection* connection) override
  {
    std::cout << "connected" << std::endl;
  }

protected:
  virtual uint16_t onNegotiate(AMQP::TcpConnection* connection, uint16_t timeout) override
  {
    std::cout << "onNegotiate" << std::endl;
    return LibEvHandler::onNegotiate(connection, timeout);
  }
  virtual void onDetached(AMQP::TcpConnection* connection) override
  {
    std::cout << "detached" << std::endl;
  }
};

int
main()
{

  OPENSSL_init_ssl(0, NULL);
  auto* loop = EV_DEFAULT;

  MyHandler myHandler{loop};

  AMQP::Address address("amqps://user:pw.host/");

  AMQP::TcpConnection connection(&myHandler, address);

  AMQP::TcpChannel channel(&connection);

  channel.onError([](const char* message) { std::cout << "cno " << message << std::endl; });

  channel.onReady([]() { std::cout << "cyes" << std::endl; });


  channel.declareExchange("test-pls-delete-exchange", AMQP::fanout)
    .onSuccess([]() { std::cout << "yes" << std::endl; })
    .onError([](const char* message) { std::cout << "no " << message << std::endl; });

  ev_run(loop, 0);
  
  return 0;
}```

Output before that commit:

```onAttached
connected
onSecuring
onSecured
onProperties
onNegotiate
ready
cyes
yes
onHeartbeat
onHeartbeat
onHeartbeat
[..]```

And the exchange is actually created!

Output after that commit:

```onAttached
connected
onSecuring
onSecured
onProperties
onNegotiate

Process finished with exit code 0```

And the exchange is **not** created.

**Comment by audaki** [comment]
The two commits I compare:
```
git checkout e60280fb209a3fa58e9f2ea3f50eba4177e41257
git checkout cffbd346aea54ee81e29f796c048a8d79ccbf44e
```

Maybe it's also important for this bug that we're using amqps so the TLS part is active.

**Comment by RafalGoslawski** [comment]
I have reproduced the issue. TLS is a red herring, as the issue is also present without it.
Looks like a bug in the commit you've mentioned. 
https://github.com/CopernicaMarketingSoftware/AMQP-CPP/blob/master/include/amqpcpp/libev.h#L372
At the very least ev_timer_again use is incorrect, since it's a non-repeating timer. @rwols

**Comment by rwols** [comment]
If I remove these ev_timer_again calls your sample program seems to keep running @audaki

**Comment by audaki** [comment]
It'd be important for me personally that it's understood why the change was introduced, why it worked before and why it stopped working, to see if removing ev_timer_again or like before using timer.repeat_ is the correct choice, to ensure that there are no further edge cases where the library breaks.

**Comment by rwols** [comment]
The history of this change can be found in

https://github.com/CopernicaMarketingSoftware/AMQP-CPP/issues/420
https://github.com/CopernicaMarketingSoftware/AMQP-CPP/pull/421
https://github.com/CopernicaMarketingSoftware/AMQP-CPP/pull/423

**Comment by rwols** [comment]
I'm noticing now that https://github.com/CopernicaMarketingSoftware/AMQP-CPP/pull/429 quits in the onHeartbeat, so it needs more work.

**Comment by audaki** [comment]
From what I read in the documentation `ev_timer_again` sets it to the configured repeat, which obviously can't work anymore after the repeat was removed.

From the documentation it seems that the repeat methodology is the fastest and best one, so it's sad that it caused problems and I don't have that much time to dig deeper into it what happened to fix it.

![Screenshot_20210916_145332](https://user-images.githubusercontent.com/1321285/133615481-629f6315-5be0-4f32-a7d5-251b11022bad.png)

What I want to point out is that using `ev_timer_set` + `ev_timer_start` should usually be used with `ev_timer_stop` before.

**Comment by rwols** [comment]
I agree, based on what I read in https://linux.die.net/man/3/ev ev_timer_again seems like the thing we want.

**Comment by EmielBruijntjes** [comment]
I think that we solved this in 658da59a1b4397198ab371cbd25e52e8d383a57e
